File: Tests/GlasstualTests/OELReachabilityMigrationTests.m
/* *********************************************************************
 * Copyright (c) 2026 Codeux Software, LLC & respective contributors.
 * Please see Acknowledgements.pdf for additional information.
 *********************************************************************** */

#import <XCTest/XCTest.h>

#import "OELReachabilityPrivate.h"

NS_ASSUME_NONNULL_BEGIN

@interface OELReachabilityMigrationTests : XCTestCase
@end

@implementation OELReachabilityMigrationTests

- (void)testFactoryCreatesNotifier
{
	OELReachability *reachability = [OELReachability reachabilityForInternetConnection];

	XCTAssertNotNil(reachability);
	XCTAssertFalse(reachability.isReachable);
}

- (void)testFirstPathSeedsWithoutEvent
{
	BOOL currentlyReachable = NO;
	BOOL receivedInitialPath = NO;

	NSInteger event = [OELReachability evaluatePathChange:YES
									   currentlyReachable:&currentlyReachable
									  receivedInitialPath:&receivedInitialPath];

	XCTAssertEqual(event, 0);
	XCTAssertTrue(currentlyReachable);
	XCTAssertTrue(receivedInitialPath);
}

- (void)testUnchangedPathProducesNoEvent
{
	BOOL currentlyReachable = YES;
	BOOL receivedInitialPath = YES;

	NSInteger event = [OELReachability evaluatePathChange:YES
									   currentlyReachable:&currentlyReachable
									  receivedInitialPath:&receivedInitialPath];

	XCTAssertEqual(event, 0);
	XCTAssertTrue(currentlyReachable);
}

- (void)testReachabilityTransitionsEmitExpectedEvents
{
	BOOL currentlyReachable = YES;
	BOOL receivedInitialPath = YES;

	NSInteger becameUnreachable = [OELReachability evaluatePathChange:NO
												   currentlyReachable:&currentlyReachable
												  receivedInitialPath:&receivedInitialPath];
	XCTAssertEqual(becameUnreachable, 2);
	XCTAssertFalse(currentlyReachable);

	NSInteger becameReachable = [OELReachability evaluatePathChange:YES
												 currentlyReachable:&currentlyReachable
												receivedInitialPath:&receivedInitialPath];
	XCTAssertEqual(becameReachable, 1);
	XCTAssertTrue(currentlyReachable);
}

- (void)testStartAndStopNotifierRoundTrip
{
	OELReachability *reachability = [OELReachability reachabilityForInternetConnection];

	XCTAssertTrue([reachability startNotifier]);
	[reachability stopNotifier];
	XCTAssertTrue([reachability startNotifier]);
	[reachability stopNotifier];
}

@end

NS_ASSUME_NONNULL_END

File: Tests/GlasstualTests/OELReachabilityMigrationTests.swift
@testable import Glasstual
import XCTest

/** *********************************************************************
 * Copyright (c) 2026 Codeux Software, LLC & respective contributors.
 * Please see Acknowledgements.pdf for additional information.
 *********************************************************************** */
@MainActor
class OELReachabilityMigrationTests: XCTestCase {
	func testFactoryCreatesNotifier() {
		let reachability = Reachability.reachabilityForInternetConnection()

		XCTAssertNotNil(reachability)
		XCTAssertFalse(reachability.reachable)
	}

	func testFirstPathSeedsWithoutEvent() {
		var currentlyReachable = false
		var receivedInitialPath = false
		let event = Reachability.evaluatePathChange(
			reachable: true,
			currentlyReachable: &currentlyReachable,
			receivedInitialPath: &receivedInitialPath
		)

		XCTAssertEqual(event, .none)
		XCTAssertTrue(currentlyReachable)
		XCTAssertTrue(receivedInitialPath)
	}

	func testUnchangedPathProducesNoEvent() {
		var currentlyReachable = true
		var receivedInitialPath = true
		let event = Reachability.evaluatePathChange(
			reachable: true,
			currentlyReachable: &currentlyReachable,
			receivedInitialPath: &receivedInitialPath
		)

		XCTAssertEqual(event, .none)
		XCTAssertTrue(currentlyReachable)
	}

	func testReachabilityTransitionsEmitExpectedEvents() {
		var currentlyReachable = true
		var receivedInitialPath = true
		let becameUnreachable = Reachability.evaluatePathChange(
			reachable: false,
			currentlyReachable: &currentlyReachable,
			receivedInitialPath: &receivedInitialPath
		)

		XCTAssertEqual(becameUnreachable, .becameUnreachable)
		XCTAssertFalse(currentlyReachable)

		let becameReachable = Reachability.evaluatePathChange(
			reachable: true,
			currentlyReachable: &currentlyReachable,
			receivedInitialPath: &receivedInitialPath
		)

		XCTAssertEqual(becameReachable, .becameReachable)
		XCTAssertTrue(currentlyReachable)
	}

	func testStartAndStopNotifierRoundTrip() {
		let reachability = Reachability.reachabilityForInternetConnection()

		XCTAssertTrue(reachability.startNotifier())

		reachability.stopNotifier()

		XCTAssertTrue(reachability.startNotifier())

		reachability.stopNotifier()
	}
}
